Add tests for the examples

MatthieuDartiailh opened the pull request.

frmdstryr commented on enaml/application.py:
```diff
@@ -261,6 +261,14 @@ def stop(self):
+    def process_events(self):
+        """ Process the events received by the application.
+
+        This allows to manually run teh application event loop.
```
Typo  in 'the'

MatthieuDartiailh replied:
Good catch I will fix it

sccolbert replied:
s/allows/allows the user

sccolbert commented on enaml/qt/qt_application.py:
```diff
@@ -56,6 +56,17 @@ def stop(self):
+    def process_events(self):
+        """ Process the events received by the application.
+
+        This allows to manually run teh application event loop.
```
Same typos

sccolbert commented on enaml/testing/utils.py:
```diff
@@ -0,0 +1,163 @@
+        Type of the window which should be returned.
+
+    """
+    sleep(0.1)
```
why is the sleep needed?

MatthieuDartiailh replied:
I think I introduced this sleep because if calling this function right after showing a widget, it would be displayed before being returned which sometimes could cause issue.

sccolbert commented on enaml/testing/utils.py:
```diff
@@ -0,0 +1,163 @@
+    w_ = None
+    for w in Window.windows:
+        if isinstance(w, cls):
+            w_ = w
```
you can just `return w` here instead of using the temp variable.

MatthieuDartiailh replied:
Will update

sccolbert commented on enaml/testing/utils.py:
```diff
@@ -0,0 +1,163 @@
+        for window in list(Window.windows):
+            window.close()
+        app.process_events()
+        sleep(0.02)
```
why sleep?

MatthieuDartiailh replied:
In case any thread decides to create a new window after this was called but also because sometimes Qt handling (and os related actions) take some time to actually take effect, even if I am not sure why.

sccolbert commented on enaml/testing/utils.py:
```diff
@@ -0,0 +1,163 @@
+    win.close()
+    app.process_events()
+
+    close_all_windows(app)
```
Lots of `sleep` and `process_events` in this file. I wonder if a decorator or context manager would be a better approach since it would be explicit and local to the test where it is needed.

MatthieuDartiailh replied:
The last two functions are a left over, we do not need them so I will remove them. I think it make sense to turn close_all_windows into a context manager. I will do so and remove the associated fixture.

MatthieuDartiailh replied:
If windows_closing disappears, we can manually manage what app is passed to the test, which is cleaner in a sense even if we have only one backend currently. I will refactor the code to skip tests if qt is not available.

MatthieuDartiailh commented:
For me this is good to go. I wish I could add more tests but this takes just too much time and I would like to work on the Python 3 compatibility.
I reworked a bit the testing package. @sccolbert if you are not happy with it, I can move it inside the tests directory and reconsider it making it public later.

frmdstryr commented on tests/test_examples.py:
```diff
@@ -0,0 +1,227 @@
+            window.show()
+            window.send_to_front()
+            qt_app.process_events()
+            sleep(enaml_sleep)
```
This is not working on my system in some cases. I created a simple hello world example with a window and a label, the text doesn't display because it the sleep is blocking. 

I had to change it to:

```python

            while enaml_sleep>0:
                qt_app.process_events()
                sleep(0.1)
                enaml_sleep -= 0.1

```

I don't think using a blocking sleep is ever a good idea

MatthieuDartiailh replied:
This is indeed not working perfectly well. I hope to rework this soon for a better solution. I think that pytest-qt may help here.

MatthieuDartiailh commented:
I refactored the testing system to use pytest-qt and it is much nicer ! much more predictive, no more stupid sleep or random process events. @frmdstryr I would really like if you could provide some feedback, in particular window should now be properly displayed. @sccolbert do you agree to include the testing package in enaml or should I move it to the tests folder ?

MatthieuDartiailh commented:
The last commit simply add window and popup cleaning to the bot fixture (so automatic teardown) which I think makes sense.

MatthieuDartiailh commented:
I just took the time to port all the tests of https://github.com/Ecpy/ecpy (will open a PR soon) to use the same functions and found that the handling of dialogs was not optimal. This is now fixed.

frmdstryr commented:
Ok finally tried this out, it's working fine. Also ran with `--enaml-sleep 0.5` and everything displayed as it should.  

Only comments I have is maybe add some docs on how to setup and run (or just say see travis.yml). Also, since it's both for testing and qt dependent I'd suggest moving the testing package into the `tests` folder. 

To use in a venv (on linux)
1. I had to `sudo apt install python-qt` and manually link PyQt4 and sip [as shown here](https://stackoverflow.com/a/28850104/2362877) to use in a venv  
2. Activate the venv
3. Ran `pip install atom ply kiwisolver pytest pytest-qt pytest-catchlog pytest-cov pytest-timeout`
4. Did `python setup.py build_ext --inplace`
5. Also had to add '.' to the `sys.path` in the `conftest.py` for whatever reason it wasn't there.

```
(venv) jrm@xps:~/Workspace/frmdstryr/enaml$ py.test tests --cov enaml
======================================================= test session starts ========================================================
platform linux2 -- Python 2.7.12, pytest-3.2.2, py-1.4.34, pluggy-0.4.0
PyQt4 4.12 -- Qt runtime 4.8.7 -- Qt compiled 4.8.7
rootdir: /home/jrm/Workspace/frmdstryr/enaml, inifile:
plugins: timeout-1.2.0, qt-2.2.0, cov-2.5.1, catchlog-1.2.2
collected 185 items                                                                                                                 

tests/test_alias.py ....................................
tests/test_examples.py ....s.................F................................s...........s...
tests/test_q_resource_helpers.py .
tests/test_stylesheet.py ...........
tests/test_template.py ..................................................................

---------- coverage: platform linux2, python 2.7.12-final-0 ----------
Name                                      Stmts   Miss  Cover
-------------------------------------------------------------
enaml/__init__.py                            22      9    59%
enaml/applib/__init__.py                      0      0   100%
enaml/applib/live_editor_model.py            93     93     0%
enaml/application.py                        144     61    58%
enaml/colors.py                             142     85    40%
enaml/core/__init__.py                        0      0   100%
enaml/core/api.py                             6      0   100%
enaml/core/block_compiler.py                 82      0   100%
enaml/core/byteplay.py                      524    166    68%
enaml/core/code_generator.py                129     16    88%
enaml/core/code_tracing.py                   66     13    80%
enaml/core/compiler_common.py               252     21    92%
enaml/core/compiler_helpers.py              257     40    84%
enaml/core/compiler_nodes.py                154      4    97%
enaml/core/conditional.py                    31      4    87%
enaml/core/declarative.py                    41      0   100%
enaml/core/declarative_function.py           44     13    70%
enaml/core/declarative_meta.py               37      0   100%
enaml/core/dynamic_template.py               64     16    75%
enaml/core/enaml_ast.py                     106      5    95%
enaml/core/enaml_compiler.py                 67      2    97%
enaml/core/enamldef_compiler.py             115      0   100%
enaml/core/enamldef_meta.py                  22      6    73%
enaml/core/expression_engine.py              72     14    81%
enaml/core/import_hooks.py                  151     39    74%
enaml/core/include.py                        76     53    30%
enaml/core/lexer.py                         341     54    84%
enaml/core/looper.py                         62      9    85%
enaml/core/object.py                        148     39    74%
enaml/core/operators.py                      78     10    87%
enaml/core/parse_tab/__init__.py              0      0   100%
enaml/core/parse_tab/lextab.py                9      1    89%
enaml/core/parse_tab/parsetab.py             18      0   100%
enaml/core/parser.py                       2185    742    66%
enaml/core/pattern.py                        25      5    80%
enaml/core/standard_handlers.py              46      7    85%
enaml/core/standard_inverter.py              18     10    44%
enaml/core/standard_tracer.py                58      6    90%
enaml/core/template.py                       92      9    90%
enaml/core/template_.py                       2      0   100%
enaml/core/template_compiler.py             154      7    95%
enaml/drag_drop.py                           39     12    69%
enaml/fonts.py                               93     24    74%
enaml/icon.py                                 9      0   100%
enaml/image.py                               10      0   100%
enaml/layout/__init__.py                      0      0   100%
enaml/layout/api.py                           3      0   100%
enaml/layout/box_helper.py                   14      0   100%
enaml/layout/constrainable.py                58      4    93%
enaml/layout/constraint_helper.py            17      2    88%
enaml/layout/dock_layout.py                 244     61    75%
enaml/layout/factory_helper.py               21      3    86%
enaml/layout/geometry.py                    172     55    68%
enaml/layout/grid_helper.py                 131      4    97%
enaml/layout/layout_helpers.py               30      0   100%
enaml/layout/layout_manager.py              193     15    92%
enaml/layout/linear_box_helper.py            82      9    89%
enaml/layout/linear_symbolic.py               7      0   100%
enaml/layout/sequence_helper.py              69     10    86%
enaml/layout/spacers.py                      59     12    80%
enaml/layout/strength_member.py              14      3    79%
enaml/mime_data.py                           12      5    58%
enaml/nodevisitor.py                         24      5    79%
enaml/objectdict.py                           7      4    43%
enaml/qt/QtCore.py                           14      2    86%
enaml/qt/QtGui.py                             4      1    75%
enaml/qt/QtNetwork.py                         4      4     0%
enaml/qt/QtOpenGL.py                          4      4     0%
enaml/qt/QtScript.py                          4      4     0%
enaml/qt/QtSvg.py                             4      4     0%
enaml/qt/QtTest.py                            4      4     0%
enaml/qt/QtWebKit.py                          4      4     0%
enaml/qt/__init__.py                         27     11    59%
enaml/qt/docking/__init__.py                  0      0   100%
enaml/qt/docking/dock_manager.py            255    162    36%
enaml/qt/docking/dock_overlay.py            248    197    21%
enaml/qt/docking/dock_resources.py            9      1    89%
enaml/qt/docking/event_types.py              16      3    81%
enaml/qt/docking/layout_builder.py          311    210    32%
enaml/qt/docking/layout_handling.py         319    271    15%
enaml/qt/docking/layout_saver.py             72     53    26%
enaml/qt/docking/proximity_handler.py        70     47    33%
enaml/qt/docking/q_bitmap_button.py          95     22    77%
enaml/qt/docking/q_dock_area.py             146     42    71%
enaml/qt/docking/q_dock_bar.py              391    312    20%
enaml/qt/docking/q_dock_container.py        414    269    35%
enaml/qt/docking/q_dock_frame.py            239    185    23%
enaml/qt/docking/q_dock_frame_layout.py      38      2    95%
enaml/qt/docking/q_dock_item.py             349    148    58%
enaml/qt/docking/q_dock_splitter.py          17      6    65%
enaml/qt/docking/q_dock_tab_widget.py       195    106    46%
enaml/qt/docking/q_dock_title_bar.py        252     98    61%
enaml/qt/docking/q_dock_window.py           221    168    24%
enaml/qt/docking/q_guide_rose.py            277    175    37%
enaml/qt/docking/q_icon_widget.py            32     17    47%
enaml/qt/docking/q_text_label.py             48      0   100%
enaml/qt/docking/style_sheets.py             13      9    31%
enaml/qt/docking/utils.py                     4      0   100%
enaml/qt/docking/xbms.py                     34      0   100%
enaml/qt/focus_registry.py                   14      4    71%
enaml/qt/q_deferred_caller.py                27      1    96%
enaml/qt/q_file_dialog_helper.py              9      9     0%
enaml/qt/q_flow_layout.py                   493    224    55%
enaml/qt/q_pixmap_painter.py                 30     30     0%
enaml/qt/q_pixmap_transition.py             167    167     0%
enaml/qt/q_popup_view.py                    459    151    67%
enaml/qt/q_resource_helpers.py               63      8    87%
enaml/qt/q_single_widget_layout.py           80      8    90%
enaml/qt/q_window_base.py                    46     13    72%
enaml/qt/qt_abstract_button.py               59     15    75%
enaml/qt/qt_action.py                        68     10    85%
enaml/qt/qt_action_group.py                  74     31    58%
enaml/qt/qt_application.py                   31      7    77%
enaml/qt/qt_bounded_date.py                  25     25     0%
enaml/qt/qt_bounded_datetime.py              25     25     0%
enaml/qt/qt_bounded_time.py                  25     25     0%
enaml/qt/qt_calendar.py                      22     22     0%
enaml/qt/qt_check_box.py                      8      0   100%
enaml/qt/qt_color_dialog.py                  72     72     0%
enaml/qt/qt_combo_box.py                     43      6    86%
enaml/qt/qt_constraints_widget.py            40     13    68%
enaml/qt/qt_container.py                    233     50    79%
enaml/qt/qt_control.py                        4      0   100%
enaml/qt/qt_date_selector.py                 29     29     0%
enaml/qt/qt_datetime_selector.py             29     29     0%
enaml/qt/qt_dialog.py                        45     11    76%
enaml/qt/qt_dock_area.py                    127     21    83%
enaml/qt/qt_dock_item.py                    100     30    70%
enaml/qt/qt_dock_pane.py                    149     36    76%
enaml/qt/qt_drag_drop.py                     33     13    61%
enaml/qt/qt_dual_slider.py                  206    106    49%
enaml/qt/qt_factories.py                    178     40    78%
enaml/qt/qt_field.py                        113     30    73%
enaml/qt/qt_file_dialog.py                   28     28     0%
enaml/qt/qt_file_dialog_ex.py                92     92     0%
enaml/qt/qt_flow_area.py                     72     10    86%
enaml/qt/qt_flow_item.py                     90     11    88%
enaml/qt/qt_focus_tracker.py                 21     21     0%
enaml/qt/qt_frame.py                         22      1    95%
enaml/qt/qt_group_box.py                     55      7    87%
enaml/qt/qt_html.py                          15      0   100%
enaml/qt/qt_image_view.py                   102     25    75%
enaml/qt/qt_ipython_console.py               92     92     0%
enaml/qt/qt_label.py                         35      1    97%
enaml/qt/qt_main_window.py                  117     46    61%
enaml/qt/qt_mdi_area.py                      31     31     0%
enaml/qt/qt_mdi_window.py                    48     48     0%
enaml/qt/qt_menu.py                         111     52    53%
enaml/qt/qt_menu_bar.py                      34     13    62%
enaml/qt/qt_mime_data.py                     19      7    63%
enaml/qt/qt_mpl_canvas.py                    47     47     0%
enaml/qt/qt_multiline_field.py               55     12    78%
enaml/qt/qt_notebook.py                     221     60    73%
enaml/qt/qt_object_combo.py                  83     25    70%
enaml/qt/qt_page.py                         139     46    67%
enaml/qt/qt_popup_view.py                    63      0   100%
enaml/qt/qt_progress_bar.py                  23      0   100%
enaml/qt/qt_push_button.py                   31      4    87%
enaml/qt/qt_radio_button.py                   8      0   100%
enaml/qt/qt_raw_widget.py                     7      7     0%
enaml/qt/qt_scintilla.py                    194    194     0%
enaml/qt/qt_scroll_area.py                   90     12    87%
enaml/qt/qt_separator.py                     42      2    95%
enaml/qt/qt_slider.py                        54      5    91%
enaml/qt/qt_spin_box.py                      53      5    91%
enaml/qt/qt_split_item.py                    62     12    81%
enaml/qt/qt_splitter.py                      79     23    71%
enaml/qt/qt_stack.py                        155    155     0%
enaml/qt/qt_stack_item.py                    37     37     0%
enaml/qt/qt_status_bar.py                    60     31    48%
enaml/qt/qt_status_item.py                   33     12    64%
enaml/qt/qt_time_selector.py                 26     26     0%
enaml/qt/qt_timer.py                         31     31     0%
enaml/qt/qt_tool_bar.py                     150     52    65%
enaml/qt/qt_tool_button.py                   53      7    87%
enaml/qt/qt_toolkit_dialog.py                30     30     0%
enaml/qt/qt_toolkit_object.py                40      5    88%
enaml/qt/qt_vtk_canvas.py                    41     41     0%
enaml/qt/qt_web_view.py                      20     20     0%
enaml/qt/qt_widget.py                       268    122    54%
enaml/qt/qt_window.py                       136     48    65%
enaml/qt/scintilla_lexers.py                 27     27     0%
enaml/qt/scintilla_tokens.py                  9      9     0%
enaml/qt/styleutil.py                       190     24    87%
enaml/runner.py                              45     45     0%
enaml/scintilla/__init__.py                   0      0   100%
enaml/scintilla/api.py                        1      1     0%
enaml/scintilla/idle_theme.py                 3      3     0%
enaml/scintilla/mono_font.py                  6      6     0%
enaml/scintilla/scintilla.py                 48     48     0%
enaml/stdlib/__init__.py                      0      0   100%
enaml/stdlib/slider_transform.py             65      8    88%
enaml/styling.py                            275     67    76%
enaml/testing/__init__.py                     0      0   100%
enaml/testing/fixtures.py                    29      4    86%
enaml/testing/utils.py                      107     13    88%
enaml/validator.py                           46     20    57%
enaml/version.py                              8      8     0%
enaml/widgets/__init__.py                     0      0   100%
enaml/widgets/abstract_button.py             29      5    83%
enaml/widgets/action.py                      39      9    77%
enaml/widgets/action_group.py                21      4    81%
enaml/widgets/api.py                         62      0   100%
enaml/widgets/bounded_date.py                31     13    58%
enaml/widgets/bounded_datetime.py            31     13    58%
enaml/widgets/bounded_time.py                31     13    58%
enaml/widgets/calendar.py                     6      0   100%
enaml/widgets/check_box.py                    7      0   100%
enaml/widgets/color_dialog.py                55     29    47%
enaml/widgets/combo_box.py                   29      5    83%
enaml/widgets/constraints_widget.py          28      5    82%
enaml/widgets/container.py                   39      6    85%
enaml/widgets/control.py                      6      0   100%
enaml/widgets/date_selector.py               16      3    81%
enaml/widgets/datetime_selector.py           16      3    81%
enaml/widgets/dialog.py                      29      5    83%
enaml/widgets/dock_area.py                   89     27    70%
enaml/widgets/dock_events.py                 23      0   100%
enaml/widgets/dock_item.py                   48     12    75%
enaml/widgets/dock_pane.py                   46     11    76%
enaml/widgets/dual_slider.py                 69     27    61%
enaml/widgets/field.py                       39     11    72%
enaml/widgets/file_dialog.py                 49     23    53%
enaml/widgets/file_dialog_ex.py              78     46    41%
enaml/widgets/flow_area.py                   32      6    81%
enaml/widgets/flow_item.py                   27      4    85%
enaml/widgets/focus_tracker.py                9      0   100%
enaml/widgets/form.py                        57      1    98%
enaml/widgets/frame.py                       17      1    94%
enaml/widgets/group_box.py                   18      3    83%
enaml/widgets/h_group.py                     21      0   100%
enaml/widgets/html.py                        14      1    93%
enaml/widgets/image_view.py                  24      4    83%
enaml/widgets/ipython_console.py             22      7    68%
enaml/widgets/label.py                       20      3    85%
enaml/widgets/main_window.py                 22      0   100%
enaml/widgets/mdi_area.py                    13      1    92%
enaml/widgets/mdi_window.py                  17      5    71%
enaml/widgets/menu.py                        39     15    62%
enaml/widgets/menu_bar.py                     9      1    89%
enaml/widgets/mpl_canvas.py                  20      5    75%
enaml/widgets/multiline_field.py             31     10    68%
enaml/widgets/notebook.py                    32      6    81%
enaml/widgets/object_combo.py                28      4    86%
enaml/widgets/page.py                        32      7    78%
enaml/widgets/popup_view.py                  83     19    77%
enaml/widgets/progress_bar.py                51     15    71%
enaml/widgets/push_button.py                 17      3    82%
enaml/widgets/radio_button.py                 7      0   100%
enaml/widgets/raw_widget.py                  13      4    69%
enaml/widgets/scroll_area.py                 27      3    89%
enaml/widgets/separator.py                   37     12    68%
enaml/widgets/slider.py                      65     21    68%
enaml/widgets/spin_box.py                    49     17    65%
enaml/widgets/split_item.py                  21      2    90%
enaml/widgets/splitter.py                    20      2    90%
enaml/widgets/stack.py                       27      5    81%
enaml/widgets/stack_item.py                  11      3    73%
enaml/widgets/status_bar.py                  25      8    68%
enaml/widgets/status_item.py                 20      2    90%
enaml/widgets/time_selector.py               13      2    85%
enaml/widgets/timer.py                       32     13    59%
enaml/widgets/tool_bar.py                    51     17    67%
enaml/widgets/tool_button.py                 18      3    83%
enaml/widgets/toolkit_dialog.py              69     40    42%
enaml/widgets/toolkit_object.py              72      4    94%
enaml/widgets/v_group.py                     18      0   100%
enaml/widgets/vtk_canvas.py                  25      8    68%
enaml/widgets/web_view.py                    18      3    83%
enaml/widgets/widget.py                     109     43    61%
enaml/widgets/window.py                     150     61    59%
enaml/workbench/__init__.py                   0      0   100%
enaml/workbench/api.py                        5      5     0%
enaml/workbench/core/__init__.py              0      0   100%
enaml/workbench/core/api.py                   1      1     0%
enaml/workbench/core/command.py               6      6     0%
enaml/workbench/core/core_plugin.py          75     75     0%
enaml/workbench/core/execution_event.py       8      8     0%
enaml/workbench/extension.py                 23     23     0%
enaml/workbench/extension_point.py           13     13     0%
enaml/workbench/plugin.py                    10     10     0%
enaml/workbench/plugin_manifest.py           19     19     0%
enaml/workbench/ui/__init__.py                0      0   100%
enaml/workbench/ui/action_item.py            19     19     0%
enaml/workbench/ui/api.py                     7      7     0%
enaml/workbench/ui/autostart.py               4      4     0%
enaml/workbench/ui/branding.py                6      6     0%
enaml/workbench/ui/item_group.py              7      7     0%
enaml/workbench/ui/menu_helper.py           168    168     0%
enaml/workbench/ui/menu_item.py              13     13     0%
enaml/workbench/ui/ui_plugin.py             231    231     0%
enaml/workbench/ui/ui_workbench.py           14     14     0%
enaml/workbench/ui/window_model.py            8      8     0%
enaml/workbench/ui/workspace.py              12     12     0%
enaml/workbench/workbench.py                125    125     0%
-------------------------------------------------------------
TOTAL                                     21917   8866    60%


============================================================= FAILURES =============================================================
______________________________________ test_examples[layout/advanced/button_ring.enaml-None] _______________________________________

enaml_qtbot = <pytestqt.qtbot.QtBot object at 0x7f38ffe0bf10>, enaml_sleep = 0, path = 'layout/advanced/button_ring.enaml'
handler = None

    @pytest.mark.parametrize("path, handler",
                             [('aliases/chained_attribute_alias.enaml', None),
                              ('aliases/chained_widget_alias.enaml', None),
                              ('aliases/simple_attribute_alias.enaml', None),
                              ('aliases/simple_widget_alias.enaml', None),
                              pytest.param('applib/live_editor.enaml', None,
                                           marks=pytest.mark.skipif(
                                               not SCINTILLA_AVAILABLE,
                                               reason='Requires scintilla')),
                              ('dynamic/conditional.enaml', None),
                              ('dynamic/fields.enaml', None),
                              ('dynamic/looper.enaml', None),
                              ('dynamic/notebook_pages.enaml', None),
                              ('functions/declare_function.enaml', None),
                              ('functions/observe_model_signal.enaml', None),
                              ('functions/override_function.enaml', None),
                              ('layout/basic/align_offset.enaml', None),
                              ('layout/basic/align.enaml', None),
                              ('layout/basic/grid.enaml', None),
                              ('layout/basic/hbox_equal_widths.enaml', None),
                              ('layout/basic/hbox_spacing.enaml', None),
                              ('layout/basic/hbox.enaml', None),
                              ('layout/basic/horizontal.enaml', None),
                              ('layout/basic/linear_relations.enaml', None),
                              ('layout/basic/vbox.enaml', None),
                              ('layout/basic/vertical.enaml', None),
                              ('layout/advanced/button_ring.enaml', None),
                              ('layout/advanced/factory_func.enaml', None),
                              ('layout/advanced/find_replace.enaml', None),
                              ('layout/advanced/fluid.enaml', None),
                              ('layout/advanced/manual_hbox.enaml', None),
                              ('layout/advanced/manual_vbox.enaml', None),
                              ('layout/advanced/nested_boxes.enaml', None),
                              ('layout/advanced/nested_containers.enaml', None),
                              ('layout/advanced/override_layout_constraints.enaml',
                               None),
                              ('stdlib/mapped_view.enaml', None),
                              ('stdlib/message_box.enaml', None),
                              ('stdlib/slider_transform.enaml', None),
                              ('stdlib/task_dialog.enaml', None),
                              ('styling/banner.enaml', None),
                              ('styling/dock_item_alerts.enaml', None),
                              ('styling/gradient_push_button.enaml', None),
                              ('templates/basic.enaml', None),
                              ('templates/advanced.enaml', None),
                              ('widgets/buttons.enaml', None),
                              ('widgets/context_menu.enaml', None),
                              ('widgets/dock_area.enaml', None),
                              ('widgets/dock_pane.enaml', None),
                              ('widgets/dual_slider.enaml', None),
                              ('widgets/file_dialog.enaml', None),
                              ('widgets/flow_area.enaml', None),
                              ('widgets/focus_traversal.enaml', None),
                              ('widgets/form_spacing.enaml', None),
                              ('widgets/form.enaml', None),
                              ('widgets/group_box.enaml', None),
                              ('widgets/h_group.enaml', None),
                              ('widgets/image_view.enaml', None),
                              ('widgets/main_window.enaml', None),
                              ('widgets/menu_bar.enaml', None),
                              pytest.param('widgets/mpl_canvas.enaml', None,
                                           marks=pytest.mark.skipif(
                                               not MATPLOTLIB_AVAILABLE,
                                               reason='Requires matplotlib')),
                              ('widgets/notebook.enaml', None),
                              ('widgets/popup_menu.enaml', None),
                              ('widgets/popup_view.enaml',
                               handle_popup_view_example),
                              ('widgets/progress_bar.enaml', None),
                              ('widgets/scroll_area.enaml', None),
                              ('widgets/slider.enaml', None),
                              ('widgets/spin_box.enaml', None),
                              ('widgets/splitter.enaml', None),
                              ('widgets/tool_bar.enaml', None),
                              ('widgets/tool_buttons.enaml', None),
                              ('widgets/v_group.enaml', None),
                              pytest.param('widgets/vtk_canvas.enaml', None,
                                           marks=pytest.mark.skipif(
                                               not VTK_AVAILABLE,
                                               reason='Requires vtk')),
                              ('widgets/window_children.enaml', None),
                              ('widgets/window_closing.enaml',
                               handle_window_closing),
                              ('widgets/window.enaml', None)])
    def test_examples(enaml_qtbot, enaml_sleep, path, handler):
        """ Test the enaml examples.
    
        """
        dir_path = os.path.abspath(os.path.split(os.path.dirname(__file__))[0])
        enaml_file = os.path.join(dir_path, 'examples', os.path.normpath(path))
    
        with open(enaml_file, 'rU') as f:
            enaml_code = f.read()
    
        # Parse and compile the Enaml source into a code object
        ast = parse(enaml_code, filename=enaml_file)
        code = EnamlCompiler.compile(ast, enaml_file)
    
        # Create a proper module in which to execute the compiled code so
        # that exceptions get reported with better meaning
        try:
            module = types.ModuleType('enaml_test')
            module.__file__ = os.path.abspath(enaml_file)
            sys.modules['enaml_test'] = module
            ns = module.__dict__
    
            # Put the directory of the Enaml file first in the path so relative
            # imports can work.
            sys.path.insert(0, os.path.abspath(os.path.dirname(enaml_file)))
            with imports():
>               exec code in ns

tests/test_examples.py:193: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    #------------------------------------------------------------------------------
    # Copyright (c) 2013, Nucleic Development Team.
    #
    # Distributed under the terms of the Modified BSD License.
    #
    # The full license is in the file COPYING.txt, distributed with this software.
    #------------------------------------------------------------------------------
    """ An example layout which is impossible with typical layout systems.
    
    This example creates a ring of 50 `PushButton` widgets with a `Label` in
    the center. The constraints shown here are generally silly in that the
    resulting layout is more-or-less useless. Nevertheless, it serves well to
    demonstrate the power and flexibility of constraints-based layout.
    
    Note that the 'gen_constraints' function is only called once, not on
    every resize as may be expected when laying out widgets manually.
    
    Note that this example also demonstrates that constraints may be defined
    on any subclass of `ConstraintsWidget`. They need not be confined to a
    `Container`.
    
    << autodoc-me >>
    """
>   import numpy as np
E   ImportError: No module named numpy

examples/layout/advanced/button_ring.enaml:24: ImportError
========================================= 1 failed, 181 passed, 3 skipped in 12.32 seconds =========================================
```

MatthieuDartiailh commented:
Thanks for testing this @frmdstryr 
From your log it appears I forgot a skipif on one test I will add it.

The rationale for having it in testing is that other project can use those tools too without duplicating them.  Another option may be to put those under  tools and installable as pytest plugin (enaml-qt), or to create a different repo.

@sccolbert any opinion on where those tools should live ?

MatthieuDartiailh commented:
I removed the testing package and moved everything back under tests. We will reconsider making a pytest plugin at some later point in the future. As this was the last questionable point here, I will merge in a week if nobody complains at it does not make sense for this small change to hold the following PRs more than necessary.
